Release: v0.12-rc1

tombh opened the pull request:
This is my first release. There isn't any documentation on how to release, and I'm having some trouble with the version number. So I'm putting this up for some feedback.

Even though I'm explicitly setting `__version__` to `0.12-rc1`, the build process always automatically decides on this version: `pygls-0.12rc2.dev0+g96bc69d.d20220704`, based on the git tag. I've noted my steps in a new `RELEASING.md` file in the root.

## Code review checklist (for code reviewer to complete)

- [ ] Pull request represents a single change (i.e. not fixing disparate/unrelated things in a single PR)
- [ ] Title summarizes what is changing
- [ ] Commit messages are meaningful (see [this][commit messages] for details)
- [ ] Tests have been included and/or updated, as appropriate
- [ ] Docstrings have been included and/or updated, as appropriate
- [ ] Standalone docs have been updated accordingly
- [ ] [CONTRIBUTORS.md][contributors] was updated, as appropriate
- [ ] Changelog has been updated, as needed (see [CHANGELOG.md][changelog])

[changelog]: https://github.com/openlawlibrary/pygls/blob/master/CHANGELOG.md
[commit messages]: https://chris.beams.io/posts/git-commit/
[contributors]: https://github.com/openlawlibrary/pygls/blob/master/CONTRIBUTORS.md

alcarney commented:
Based on [this](https://github.com/pypa/setuptools_scm#default-versioning-scheme) I guess `setuptools_scm` doesn't take `__version__` into account and that you need to do the build with a clean checkout at the exact version tag to set the version number correctly?

tombh commented:
That was it! Thanks @alcarney 🤓 The PR is ready for review now (disregarding the flakey test runs, I'm still working on #237 to address that).

I made [a real Pypi RC release](https://pypi.org/project/pygls/0.12rc1/#files), but without your latest #247. So we could make an `0.12-rc2` or just go for a `0.12` release next. 

I updated `RELEASING.md` with everything I learned. I actually simplified the CI build steps to reflect what appears to be current best practice of using `python -m build`. The version not being even remotely right was just because the CI build was doing a default shallow checkout that doesn't include tags, so I added an option to do a full checkout for the build stage.

It's worth noting that the whole process could be fully automated pretty easily as Pypi provides an API token (indeed they're deprecating HTTP Basic Auth releases). Theoretically a full release could be automatically made just by pushing a commit with a new Git tag.

alcarney commented on CHANGELOG.md:
```diff
@@ -313,7 +321,8 @@ and this project adheres to [Semantic Versioning][semver].
 
-[Unreleased]: https://github.com/openlawlibrary/pygls/compare/v0.11.3...HEAD
+[Unreleased]: https://github.com/openlawlibrary/pygls/compare/v0.12-rc1...HEAD
+[0.12-rc1]: https://github.com/openlawlibrary/pygls/compare/v0.11.3...v0.12-rc1
```
```suggestion
[Unreleased]: https://github.com/openlawlibrary/pygls/compare/v0.12-rc2...HEAD
[0.12-rc2]: https://github.com/openlawlibrary/pygls/compare/v0.11.3...v0.12-rc2
```

alcarney reviewed (dismissed):
Never done a formal review on Github before! I'm going to assume that by approving this now, it counts as "approved with suggestions" or something? :)

tombh commented:
Yeah, looks like you did it right, thanks. I'll show it to the rest of the team and see if we can get an official release.

Hey, ti'n dod o Gymru!?? I grew up in Dyfed!! I'm gonna send you an email 😃

renatav commented on CHANGELOG.md:
```diff
@@ -313,7 +321,9 @@ and this project adheres to [Semantic Versioning][semver].
 [semver]: https://semver.org/spec/v2.0.0.html
 
-[Unreleased]: https://github.com/openlawlibrary/pygls/compare/v0.11.3...HEAD
+[Unreleased]: https://github.com/openlawlibrary/pygls/compare/v0.12-rc1...HEAD
```
Duplicate line

tombh replied:
I think that's just the way the diff looks?

dimbleby replied:
no, you have two 'Unreleased' entries

tombh replied:
Ah yes! Sorry, I hadn't pulled those changes locally. I've pushed now again changing `0.12-rc2` to just `0.12`.

dimbleby replied:
you seem actually to have changed to `0.12-rc1`, but close enough I guess

tombh replied:
uff damn that's a shame. thanks for pointing that out

renatav reviewed (commented):
On another project, we have azure set up and automatically deploy to pypi after tagging. However, in that case we need to build platform specific wheels, so that automation is much more important. 
We usually wait until the release branches are reviewed and merged before tagging and uploading to pypi

tombh commented:
Ok, so when the tests go green let's make a release!

tombh commented:
@renatav So you're using Azure because Github Actions can't build for the platforms you require?

If we were to make official releases automatically I think Github Actions are enough. But I don't think automation is priority, just interesting to consider.

renatav reviewed (approved):
LGTM
